File: MPP/MPP Proiect/CSharp/networking/ServicesRpcProxy.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using log4net;
using model;
using services;

namespace networking{
    public class ServicesRpcProxy: IServices{
        private static readonly ILog Logger = LogManager.GetLogger("ServicesRpcProxy");
        private readonly string _host;
        private readonly int _port;
        private IObserver _refereeObserver;
        private NetworkStream _stream;
        private IFormatter _formatter;
        private TcpClient _connection;
        private readonly Queue<Response> _responses;
        private volatile bool _finished;
        private EventWaitHandle _waitHandle;

        public ServicesRpcProxy(string host, int port){
            _host = host;
            _port = port;
            _responses = new Queue<Response>();
        }

        private void CloseConnection(){
            _finished = true;
            try{
                _stream.Close();
                _connection.Close();
                _waitHandle.Close();
                _refereeObserver = null;
            } catch(Exception e){
                Console.WriteLine(e.StackTrace);
            }
        }

        private void SendRequest(Request request){
            try{
                _formatter.Serialize(_stream, request);
                _stream.Flush();
            } catch(Exception e){
                throw new MyException("Error sending object " + e);
            }
        }

        private Response ReadResponse(){
            Response response = null;
            try{
                _waitHandle.WaitOne();
                lock(_responses){
                    response = _responses.Dequeue();
                }
            } catch(Exception e){
                Console.WriteLine(e.StackTrace);
            }

            return response;
        }

        private void InitializeConnection(){
            try{
                _connection = new TcpClient(_host, _port);
                _stream = _connection.GetStream();
                _formatter = new BinaryFormatter();
                _finished = false;
                _waitHandle = new AutoResetEvent(false);
                StartReader();
            } catch(Exception e){
                Console.WriteLine(e.StackTrace);
            }
        }

        private void StartReader(){
            var tw = new Thread(Run);
            tw.Start();
        }

        protected virtual void Run(){
            while(!_finished){
                try{
                    var response = _formatter.Deserialize(_stream);
                    Console.WriteLine("response received " + response);
                    if(IsUpdate((Response)response)){
                        HandleUpdate((Response)response);
                    } else{
                        lock(_responses){
                            _responses.Enqueue((Response)response);
                        }

                        _waitHandle.Set();
                    }
                } catch(Exception e){
                    Console.WriteLine("Reading error " + e);
                }
            }
        }

        private void HandleUpdate(Response response){
            if(response.Type != ResponseType.TEST_ADDED) return;
            Logger.Info("Test added ");
            try{
                _refereeObserver.RefreshAll(response.Data as Test);
            } catch(MyException e){
                Logger.Error("Error handle update: " + e);
            }
        }

        private static bool IsUpdate(Response response){
            return response.Type == ResponseType.TEST_ADDED;
        }

        public Referee Login(Referee referee, IObserver client){
            InitializeConnection();
            var request = new Request.Builder().Type(RequestType.LOGIN)
                .Data(referee).Build();
            SendRequest(request);
            var response = ReadResponse();
            switch(response.Type){
                case ResponseType.OK:
                    _refereeObserver = client;
                    Logger.Info("Login successful");
                    return (Referee)response.Data;
                case ResponseType.ERROR:
                    Logger.Info("Login failed" + response.Data);
                    CloseConnection();
                    throw new MyException((string)response.Data);
                default:
                    return null;
            }
        }

        public void AddTest(Test test){
            var request = new Request.Builder().Type(RequestType.ADD_TEST)
                .Data(test).Build();
            SendRequest(request);
            var response = ReadResponse();
            if(response.Type == ResponseType.ERROR){
                Logger.Info("Add test failed" + response.Data);
                throw new MyException((string)response.Data);
            }

            Console.WriteLine("Add test successful");
            Logger.Info("Add test successful");
        }

        public IEnumerable<Participant> GetAllParticipantsSortedByName(){
            var request = new Request.Builder().Type(RequestType.GET_ALL_PARTICIPANTS)
                .Build();
            SendRequest(request);
            var response = ReadResponse();
            if(response.Type == ResponseType.ERROR){
                Logger.Info("Get all participants failed" + response.Data);
                throw new MyException((string)response.Data);
            }

            Logger.Info("Get all participants successful");
            return (IEnumerable<Participant>)response.Data;
        }

        public IEnumerable<Participant> GetAllActiveParticipants(int refereeId){
            var request = new Request.Builder()
                .Type(RequestType.GET_ALL_ACTIVE_PARTICIPANTS)
                .Data(refereeId).Build();
            SendRequest(request);
            var response = ReadResponse();
            if(response.Type == ResponseType.ERROR){
                Logger.Info("Get all active participants failed" + response.Data);
                throw new MyException((string)response.Data);
            }

            Logger.Info("Get all active participants successful");
            return (IEnumerable<Participant>)response.Data;
        }

        public void Logout(Referee referee){
            var request = new Request.Builder().Type(RequestType.LOGOUT)
                .Data(referee).Build();
            SendRequest(request);
            var response = ReadResponse();
            CloseConnection();
            if(response.Type == ResponseType.ERROR){
                Logger.Info("Logout failed" + response.Data);
                throw new MyException((string)response.Data);
            }

            Logger.Info("Logged out");
        }
    }
}

File: MPP/MPP Proiect/Java/Networking/src/main/java/rpcprotocol/ServicesRpcProxy.java
package rpcprotocol;


import model.Participant;
import model.Referee;
import model.Test;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import service.IObserver;
import service.IService;
import service.MyException;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.net.SocketException;
import java.util.ArrayList;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public
class ServicesRpcProxy implements IService{

    private final String host;
    private final int port;

    private IObserver refereeObserver;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;
    private static final Logger logger = LogManager.getLogger();

    private final BlockingQueue<Response> qresponses;
    private volatile boolean finished;

    public ServicesRpcProxy(String host, int port){
        logger.info("ServicesRpcProxy created");
        this.host = host;
        this.port = port;
        this.qresponses = new LinkedBlockingQueue<>();
    }

    public void logout(Referee referee, IObserver client) throws MyException{
        Request req =
                new Request.Builder().type(RequestType.LOGOUT).data(referee).build();
        sendRequest(req);
        Response response = readResponse();
        closeConnection();
        if(response.type() == ResponseType.ERROR){
            logger.error("Error logging out" + response.data().toString());
            String err = response.data().toString();
            throw new MyException(err);
        }else{
            logger.info("Logged out");
        }
    }

    private void closeConnection(){
        finished = true;
        try{
            input.close();
            output.close();
            connection.close();
            refereeObserver = null;
            logger.info("Closed connection");
        }catch(IOException e){
            logger.error("Error closing connection: " + e);
        }
    }

    private void sendRequest(Request request) throws MyException{
        try{
            output.writeObject(request);
            output.flush();
            logger.info("Request sent: " + request);
        }catch(IOException e){
            logger.error("Error sending object " + e);
            throw new MyException("Error sending object " + e);
        }
    }

    private Response readResponse() throws MyException{
        Response response;
        try{
            response = qresponses.take();
            logger.info("Response received: " + response);
        }catch(InterruptedException e){
            logger.error("Reading response error " + e);
            throw new MyException("Reading response error " + e);
        }
        return response;
    }

    private void initializeConnection() throws MyException{
        try{
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
            logger.info("Connection initialized");
        }catch(IOException e){
            logger.error("Error connecting to server " + e);
            throw new MyException("Error connecting to server " + e);
        }
    }

    private void startReader(){
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }

    private void handleUpdate(Response response){
        if(response.type() == ResponseType.TEST_ADDED){
            logger.info("Test added.");
            refereeObserver.refreshAll((Test) response.data());
        }
    }

    private boolean isUpdate(Response response){
        return response.type() == ResponseType.TEST_ADDED;
    }


    @Override
    public Referee login(Referee referee, IObserver refereeObserver) throws MyException{
        initializeConnection();
        Request req =
                new Request.Builder().type(RequestType.LOGIN).data(referee).build();
        sendRequest(req);
        Response response = readResponse();
        if(response.type() == ResponseType.OK){
            this.refereeObserver = refereeObserver;
            logger.info("Logged in " + referee.toString());
            return (Referee) response.data();
        }
        if(response.type() == ResponseType.ERROR){
            logger.error("Error logging in" + response.data().toString());
            String err = response.data().toString();
            closeConnection();
            throw new MyException(err);
        }
        return null;
    }

    @Override
    public Iterable<Participant> getAllParticipants() throws MyException{
        Request req = new Request.Builder().type(RequestType.GET_ALL_PARTICIPANTS)
                .data(null).build();
        sendRequest(req);
        Response response = readResponse();
        if(response.type() == ResponseType.ERROR){
            logger.error("Error getting participants" + response.data().toString());
            String err = response.data().toString();
            throw new MyException(err);
        }else{
            logger.info("Got participants");
        }
        return (ArrayList<Participant>) response.data();
    }

    @Override
    public Iterable<Participant> getAllActiveParticipants(int id) throws MyException{
        Request req = new Request.Builder().type(RequestType.GET_ALL_ACTIVE_PARTICIPANTS)
                .data(id).build();
        sendRequest(req);
        Response response = readResponse();
        if(response.type() == ResponseType.ERROR){
            logger.error(
                    "Error getting active participants" + response.data().toString());
            String err = response.data().toString();
            throw new MyException(err);
        }else{
            logger.info("Got active participants");
        }
        return (ArrayList<Participant>) response.data();
    }

    @Override
    public void addTest(Test test) throws MyException{
        Request req =
                new Request.Builder().type(RequestType.ADD_TEST).data(test).build();
        sendRequest(req);
        Response response = readResponse();
        if(response.type() == ResponseType.ERROR){
            logger.error("Error adding test" + response.data().toString());
            String err = response.data().toString();
            throw new MyException(err);
        }else{
            logger.info("Test added");
        }
    }

    private
    class ReaderThread implements Runnable{
        public void run(){
            while(!finished){
                try{
                    Object response = input.readObject();
                    logger.info("response received " + response);
                    if(isUpdate((Response) response)){
                        handleUpdate((Response) response);
                    }else{
                        try{
                            qresponses.put((Response) response);
                        }catch(InterruptedException e){
                            logger.error("Queue putting response error: " + e);
                        }
                    }
                }catch(IOException | ClassNotFoundException e){
                    if(e instanceof SocketException)
                        logger.info("Socket closed: " + e);
                    else
                        logger.error("Reading error: " + e);
                }
            }
        }
    }
}
